File: overlay/widgets/inputs.py
"""
Input telemetry -- a scrolling throttle/brake/clutch trace.

A horizontal panel (styled like the dash/table cards) showing, left to right:

  * a vertical "TELEMETRY" tab with a colored accent bar,
  * a rolling line graph of the driver inputs over the last few seconds
    (throttle green, brake red, optional clutch blue), newest on the right,
  * thin vertical bars with the current value of each input (0..100), and
  * a round medallion with the current gear and speed.

Everything is configurable from config.CFG["inputs"]: which channels show, the
history length, which sections are visible, colors and a per-widget text scale.

Expected data dict (all optional; missing values render as 0 / "--"):
    throttle, brake, clutch   0..1 pedal travel
    gear                      gear number ("R"/"N"/1..)
    speed_ms                  speed in m/s (converted to mph/kph)
"""

from __future__ import annotations

from collections import deque

from PyQt6.QtCore import QElapsedTimer, QPointF, QRectF, Qt
from PyQt6.QtGui import (QColor, QFont, QLinearGradient, QPainter, QPainterPath,
                         QPen)
from PyQt6.QtWidgets import QSizePolicy, QWidget

from .. import config


def _gear_str(g) -> str:
    if g is None:
        return "N"
    g = int(g)
    return "R" if g < 0 else ("N" if g == 0 else str(g))


class InputTraceWidget(QWidget):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.data: dict = {}
        # Rolling history of (t_seconds, throttle, brake, clutch).
        self._hist: deque = deque()
        self._clock = QElapsedTimer()
        self._clock.start()
        self._font_cache: dict = {}
        self.setMinimumSize(360, 120)
        self.setSizePolicy(QSizePolicy.Policy.Expanding, QSizePolicy.Policy.Expanding)

    # -- data --------------------------------------------------------------
    def set_data(self, data: dict) -> None:
        self.data = data or {}
        t = self._clock.elapsed() / 1000.0

        def frac(key, default=0.0):
            v = self.data.get(key)
            return max(0.0, min(1.0, float(v))) if isinstance(v, (int, float)) else default

        # Sample: (t, throttle, brake, clutch, steering, abs). Steering is already
        # normalized 0..1 (0.5 = centered); abs is 1.0 while ABS is engaged.
        abs_on = 1.0 if self.data.get("abs_active") else 0.0
        self._hist.append((t, frac("throttle"), frac("brake"), frac("clutch"),
                           frac("steer", 0.5), abs_on))
        window = float(self._cfg().get("history_seconds", 6.0) or 6.0)
        cutoff = t - window
        while len(self._hist) > 2 and self._hist[0][0] < cutoff:
            self._hist.popleft()
        self.update()

    # -- helpers -----------------------------------------------------------
    def _cfg(self) -> dict:
        return config.CFG["inputs"]

    def _col(self, key: str) -> QColor:
        cols = self._cfg()["colors"]
        return config.qcolor(cols.get(key, "#ff00ff"))

    def _font(self, px: float, bold: bool = True) -> QFont:
        fam = config.CFG.get("font_family", "Segoe UI")
        pxi = max(6, int(round(px * config.text_scale_for("inputs"))))
        key = (fam, pxi, bold)
        f = self._font_cache.get(key)
        if f is None:
            f = QFont(fam)
            f.setStyleHint(QFont.StyleHint.SansSerif)
            f.setPixelSize(pxi)
            f.setBold(bold)
            if len(self._font_cache) > 128:
                self._font_cache.clear()
            self._font_cache[key] = f
        return f

    def _bar_channels(self, c) -> list:
        """[(data_index, color_key)] for the value bars (no steering bar)."""
        out = []
        for di, flag, colk in ((1, "show_throttle", "throttle"),
                               (2, "show_brake", "brake"),
                               (3, "show_clutch", "clutch")):
            if c.get(flag, di < 3):  # throttle + brake default on, clutch off
                out.append((di, colk))
        return out

    def _brake_threshold(self, c) -> float:
        """Threshold as a 0..1 fraction (config stores it as a 0..100 percent)."""
        if not c.get("show_brake_threshold", False):
            return 0.0
        return max(0.0, min(1.0, float(c.get("brake_threshold", 0) or 0) / 100.0))

    def _brake_color(self, value, abs_on, thr, c) -> QColor:
        """Brake color: ABS yellow > over-threshold orange > normal red."""
        if abs_on:
            return self._col("brake_abs")
        if thr > 0 and value > thr:
            return self._col("brake_over")
        return self._col("brake")

    # -- paint -------------------------------------------------------------
    def paintEvent(self, event) -> None:  # noqa: N802
        p = QPainter(self)
        p.setRenderHint(QPainter.RenderHint.Antialiasing)
        w, h = float(self.width()), float(self.height())
        c = self._cfg()

        # Panel card matching the other widgets.
        radius = max(8.0, min(w, h) * 0.12)
        grad = QLinearGradient(0, 0, 0, h)
        grad.setColorAt(0.0, self._col("bg_top"))
        grad.setColorAt(1.0, self._col("bg_bottom"))
        p.setBrush(grad)
        p.setPen(QPen(self._col("panel_border"), 1.2))
        p.drawRoundedRect(QRectF(0.5, 0.5, w - 1, h - 1), radius, radius)

        pad = max(6.0, h * 0.08)
        left = pad
        right = w - pad
        gap = max(6.0, h * 0.06)
        chans = self._bar_channels(c)

        if c.get("show_label", True):
            left = self._draw_label(p, left, pad, h, c) + gap
        if c.get("show_gauge", True):
            gd = h - 2 * pad
            self._draw_gauge(p, QRectF(right - gd, pad, gd, gd), c)
            right -= gd + gap
        if c.get("show_bars", True) and chans:
            bw = max(14.0, h * 0.13)
            bgap = max(10.0, h * 0.12)
            block = len(chans) * bw + (len(chans) - 1) * bgap
            self._draw_bars(p, QRectF(right - block, pad, block, h - 2 * pad),
                            bw, bgap, chans, c)
            right -= block + gap
        if c.get("show_graph", True):
            self._draw_graph(p, QRectF(left, pad, max(1.0, right - left),
                                       h - 2 * pad), c)

    # -- sections ----------------------------------------------------------
    def _draw_label(self, p, x, pad, h, c) -> float:
        """Accent bar + rotated title; returns the x just past the label tab."""
        bar_w = max(3.0, h * 0.035)
        p.setPen(Qt.PenStyle.NoPen)
        p.setBrush(self._col("accent"))
        p.drawRoundedRect(QRectF(x, pad, bar_w, h - 2 * pad), bar_w / 2, bar_w / 2)
        text = str(c.get("label_text", "TELEMETRY"))
        p.setFont(self._font(max(8.0, h * 0.12), bold=True))
        p.setPen(self._col("label"))
        tab_w = max(14.0, h * 0.20)
        cx = x + bar_w + tab_w / 2
        length = h * 0.92
        p.save()
        p.translate(cx, h / 2)
        p.rotate(-90)
        p.drawText(QRectF(-length / 2, -tab_w / 2, length, tab_w),
                   Qt.AlignmentFlag.AlignCenter, text)
        p.restore()
        return x + bar_w + tab_w

    def _draw_graph(self, p, rect: QRectF, c) -> None:
        # Backing well + faint gridlines at 0 / 50 / 100%.
        p.setPen(Qt.PenStyle.NoPen)
        p.setBrush(self._col("graph_bg"))
        p.drawRoundedRect(rect, 6, 6)
        p.setPen(QPen(self._col("grid"), 1.0))
        for fr in (0.0, 0.5, 1.0):
            y = rect.bottom() - fr * rect.height()
            p.drawLine(QPointF(rect.left(), y), QPointF(rect.right(), y))

        if len(self._hist) < 2:
            return
        window = float(c.get("history_seconds", 6.0) or 6.0)
        now = self._hist[-1][0]
        lw = float(c.get("line_width", 2.4) or 2.4)

        def to_pt(t, frac):
            x = rect.right() - min(1.0, (now - t) / window) * rect.width()
            y = rect.bottom() - frac * rect.height()
            return QPointF(x, y)

        p.setClipRect(rect)
        # Steady-color channels first; brake (which can change color) on top.
        if c.get("show_throttle", True):
            self._line(p, 1, self._col("throttle"), to_pt, lw)
        if c.get("show_clutch", False):
            self._line(p, 3, self._col("clutch"), to_pt, lw)
        if c.get("show_steering", False):
            self._line(p, 4, self._col("steering"), to_pt, lw)

        thr = self._brake_threshold(c)
        if thr > 0:
            y = rect.bottom() - thr * rect.height()
            pen = QPen(self._col("threshold"), 1.4)
            pen.setStyle(Qt.PenStyle.DashLine)
            p.setPen(pen)
            p.drawLine(QPointF(rect.left(), y), QPointF(rect.right(), y))

        if c.get("show_brake", True):
            self._brake_line(p, to_pt, lw, thr, c)
        p.setClipping(False)

    def _line(self, p, di, color, to_pt, lw) -> None:
        path = QPainterPath()
        for j, sample in enumerate(self._hist):
            pt = to_pt(sample[0], sample[di])
            path.moveTo(pt) if j == 0 else path.lineTo(pt)
        pen = QPen(color, lw)
        pen.setJoinStyle(Qt.PenJoinStyle.RoundJoin)
        pen.setCapStyle(Qt.PenCapStyle.RoundCap)
        p.setPen(pen)
        p.setBrush(Qt.BrushStyle.NoBrush)
        p.drawPath(path)

    def _brake_line(self, p, to_pt, lw, thr, c) -> None:
        # Drawn segment by segment so each piece can take the ABS / over-threshold
        # / normal color based on that sample's state.
        prev = None
        for sample in self._hist:
            cur = to_pt(sample[0], sample[2])
            if prev is not None:
                pen = QPen(self._brake_color(sample[2], sample[5] > 0.5, thr, c), lw)
                pen.setJoinStyle(Qt.PenJoinStyle.RoundJoin)
                pen.setCapStyle(Qt.PenCapStyle.RoundCap)
                p.setPen(pen)
                p.drawLine(prev, cur)
            prev = cur

    def _draw_bars(self, p, rect: QRectF, bw, bgap, chans, c) -> None:
        latest = self._hist[-1] if self._hist else (0, 0.0, 0.0, 0.0, 0.5, 0.0)
        abs_on = latest[5] > 0.5
        thr = self._brake_threshold(c)
        label_h = max(10.0, rect.height() * 0.16)
        track_top = rect.top() + label_h
        track_h = rect.height() - label_h
        font = self._font(max(8.0, rect.height() * 0.14), bold=True)
        x = rect.left()
        for di, colk in chans:
            val = max(0.0, min(1.0, latest[di]))
            fill = (self._brake_color(val, abs_on, thr, c) if di == 2
                    else self._col(colk))
            track = QRectF(x, track_top, bw, track_h)
            r = bw * 0.4
            p.setPen(Qt.PenStyle.NoPen)
            p.setBrush(self._col("bar_track"))
            p.drawRoundedRect(track, r, r)
            fh = val * track_h
            if fh > 0:
                p.setBrush(fill)
                p.drawRoundedRect(QRectF(x, track_top + track_h - fh, bw, fh), r, r)
            p.setFont(font)
            p.setPen(self._col("text"))
            p.drawText(QRectF(x - bgap / 2, rect.top(), bw + bgap, label_h),
                       Qt.AlignmentFlag.AlignCenter, f"{val * 100:.0f}")
            x += bw + bgap

    def _draw_gauge(self, p, rect: QRectF, c) -> None:
        d = self.data or {}
        cx, cy = rect.center().x(), rect.center().y()
        rad = rect.width() / 2

        p.setBrush(self._col("gauge_bg"))
        p.setPen(QPen(self._col("gauge_ring"), max(2.0, rad * 0.10)))
        p.drawEllipse(rect.adjusted(rad * 0.06, rad * 0.06, -rad * 0.06, -rad * 0.06))

        # Small accent marker at the top of the ring (decorative, matches style).
        p.setPen(Qt.PenStyle.NoPen)
        p.setBrush(self._col("text"))
        m = rad * 0.13
        marker = QPainterPath()
        top = QPointF(cx + rad * 0.62, rect.top() + rad * 0.30)
        marker.moveTo(top.x(), top.y() - m)
        marker.lineTo(top.x() + m, top.y())
        marker.lineTo(top.x(), top.y() + m)
        marker.lineTo(top.x() - m, top.y())
        marker.closeSubpath()
        p.drawPath(marker)

        # Gear (big), then unit + speed stacked below it.
        p.setPen(self._col("text"))
        p.setFont(self._font(rad * 0.95, bold=True))
        p.drawText(QRectF(cx - rad, cy - rad * 0.78, rad * 2, rad * 1.15),
                   Qt.AlignmentFlag.AlignCenter, _gear_str(d.get("gear")))
        speed = config.conv_speed(d.get("speed_ms"))
        p.setFont(self._font(rad * 0.26, bold=True))
        p.setPen(self._col("muted"))
        p.drawText(QRectF(cx - rad, cy + rad * 0.18, rad * 2, rad * 0.34),
                   Qt.AlignmentFlag.AlignCenter, config.speed_unit())
        p.setPen(self._col("text"))
        p.setFont(self._font(rad * 0.34, bold=True))
        p.drawText(QRectF(cx - rad, cy + rad * 0.46, rad * 2, rad * 0.40),
                   Qt.AlignmentFlag.AlignCenter,
                   f"{speed:.0f}" if speed is not None else "--")

File: crates/gridglance-overlay/src/widgets/inputs.rs
//! Scrolling input telemetry (Python `inputs.py` parity).

use super::WidgetCtx;
use crate::chrome::{anim_dt, ease, full_rect, label, panel_card, still_easing};
use egui::{Align2, Color32, CornerRadius, Pos2, Rect, Shape, Stroke, Ui, Vec2};
use std::collections::VecDeque;
use std::sync::Mutex;
use std::time::Instant;

const SECTION: &str = "inputs";
const BAR_EASE_TAU: f32 = 0.07;

/// (t, thr, brk, clt, steer, abs, gear)
type Sample = (f64, f32, f32, f32, f32, f32, i32);

#[derive(Clone, Default)]
struct BarAnim {
    thr: f32,
    brk: f32,
    clt: f32,
    last_secs: f64,
}

struct Hist {
    clock: Instant,
    samples: VecDeque<Sample>,
}

impl Hist {
    fn new() -> Self {
        Self {
            clock: Instant::now(),
            samples: VecDeque::with_capacity(512),
        }
    }
}

fn hist() -> &'static Mutex<Hist> {
    use std::sync::OnceLock;
    static CELL: OnceLock<Mutex<Hist>> = OnceLock::new();
    CELL.get_or_init(|| Mutex::new(Hist::new()))
}

fn gear_str(g: i32) -> String {
    if g < 0 {
        "R".into()
    } else if g == 0 {
        "N".into()
    } else {
        g.to_string()
    }
}

fn push_sample(ctx: &WidgetCtx<'_>) {
    let f = ctx.frame;
    let cfg = ctx.cfg;
    let mut h = hist().lock().unwrap_or_else(|e| e.into_inner());
    let t = h.clock.elapsed().as_secs_f64();
    let thr = f.throttle.clamp(0.0, 1.0);
    let brk = f.brake.clamp(0.0, 1.0);
    let clt = f.clutch.clamp(0.0, 1.0);
    let steer = ((f.steering + 1.0) * 0.5).clamp(0.0, 1.0);
    let abs_on = if f.abs_active { 1.0 } else { 0.0 };
    let gear = f.gear;
    // Always append on a short cadence so the graph scrolls smoothly even when
    // pedals are held steady (otherwise the trace freezes then jumps).
    let force = match h.samples.back() {
        Some(last) => t - last.0 >= 1.0 / 30.0,
        None => true,
    };
    if let Some(last) = h.samples.back() {
        if !force
            && (last.1 - thr).abs() < 1e-4
            && (last.2 - brk).abs() < 1e-4
            && (last.3 - clt).abs() < 1e-4
            && (last.4 - steer).abs() < 1e-4
            && last.6 == gear
        {
            return;
        }
    }
    h.samples.push_back((t, thr, brk, clt, steer, abs_on, gear));
    let window = cfg.f64_key(SECTION, "history_seconds", 6.0);
    let cutoff = t - window;
    while h.samples.len() > 2 && h.samples.front().map(|s| s.0 < cutoff).unwrap_or(false) {
        h.samples.pop_front();
    }
}

pub fn paint(ui: &mut Ui, ctx: &mut WidgetCtx<'_>) {
    push_sample(ctx);
    let rect = full_rect(ui);
    panel_card(ui, ctx.cfg, SECTION, rect);
    let h = rect.height();
    let pad = if crate::chrome::is_elegant(ctx.cfg, SECTION) {
        (h * 0.06).max(5.0)
    } else {
        (h * 0.08).max(6.0)
    };
    let gap = (h * 0.06).max(6.0);
    let mut left = rect.left() + pad;
    let mut right = rect.right() - pad;

    if ctx.cfg.bool_key(SECTION, "show_label", true) && !crate::chrome::is_elegant(ctx.cfg, SECTION)
    {
        left = draw_label(ui, ctx, left, pad, h) + gap;
    }
    if ctx.cfg.bool_key(SECTION, "show_gauge", true) {
        let gd = h - 2.0 * pad;
        draw_gauge(
            ui,
            ctx,
            Rect::from_min_size(Pos2::new(right - gd, rect.top() + pad), Vec2::new(gd, gd)),
        );
        right -= gd + gap;
    }
    let chans = bar_channels(ctx);
    if ctx.cfg.bool_key(SECTION, "show_bars", true) && !chans.is_empty() {
        let bw = (h * 0.13).max(14.0);
        let bgap = (h * 0.12).max(10.0);
        let block = chans.len() as f32 * bw + (chans.len().saturating_sub(1) as f32) * bgap;
        draw_bars(
            ui,
            ctx,
            Rect::from_min_size(
                Pos2::new(right - block, rect.top() + pad),
                Vec2::new(block, h - 2.0 * pad),
            ),
            bw,
            bgap,
            &chans,
        );
        right -= block + gap;
    }
    if ctx.cfg.bool_key(SECTION, "show_graph", true) {
        draw_graph(
            ui,
            ctx,
            Rect::from_min_size(
                Pos2::new(left, rect.top() + pad),
                Vec2::new((right - left).max(1.0), h - 2.0 * pad),
            ),
        );
    }
}

fn bar_channels(ctx: &WidgetCtx<'_>) -> Vec<(usize, &'static str)> {
    let mut out = Vec::new();
    if ctx.cfg.bool_key(SECTION, "show_throttle", true) {
        out.push((1, "throttle"));
    }
    if ctx.cfg.bool_key(SECTION, "show_brake", true) {
        out.push((2, "brake"));
    }
    if ctx.cfg.bool_key(SECTION, "show_clutch", false) {
        out.push((3, "clutch"));
    }
    out
}

fn brake_threshold(ctx: &WidgetCtx<'_>) -> f32 {
    if !ctx.cfg.bool_key(SECTION, "show_brake_threshold", false) {
        return 0.0;
    }
    (ctx.cfg.f64_key(SECTION, "brake_threshold", 85.0) as f32 / 100.0).clamp(0.0, 1.0)
}

fn brake_color(ctx: &WidgetCtx<'_>, value: f32, abs_on: bool, thr: f32) -> Color32 {
    if abs_on {
        return ctx.cfg.color(SECTION, "brake_abs", "#ffd23a");
    }
    if thr > 0.0 && value > thr {
        return ctx.cfg.color(SECTION, "brake_over", "#ff7a1a");
    }
    ctx.cfg.color(SECTION, "brake", "#e23b3b")
}

fn draw_label(ui: &mut Ui, ctx: &WidgetCtx<'_>, x: f32, pad: f32, h: f32) -> f32 {
    let bar_w = (h * 0.035).max(3.0);
    let bar = Rect::from_min_size(Pos2::new(x, pad), Vec2::new(bar_w, h - 2.0 * pad));
    ui.painter().rect_filled(
        bar,
        CornerRadius::same(2),
        ctx.cfg.color(SECTION, "accent", "#e23b3b"),
    );
    let text = ctx.cfg.str_key(SECTION, "label_text", "TELEMETRY");
    let tab_w = (h * 0.20).max(14.0);
    let cx = x + bar_w + tab_w * 0.5;
    // egui has no easy rotate; draw stacked letters vertically.
    let fs = (h * 0.10).max(8.0);
    let chars: Vec<char> = text.chars().collect();
    let n = chars.len().max(1) as f32;
    let step = (h - 2.0 * pad) / n;
    for (i, ch) in chars.iter().enumerate() {
        label(
            ui,
            Pos2::new(cx, pad + step * (i as f32 + 0.5)),
            Align2::CENTER_CENTER,
            &ch.to_string(),
            fs,
            ctx.cfg.color(SECTION, "label", "#cdd3db"),
            true,
        );
    }
    x + bar_w + tab_w
}

fn draw_graph(ui: &mut Ui, ctx: &WidgetCtx<'_>, rect: Rect) {
    ui.painter().rect_filled(
        rect,
        CornerRadius::same(6),
        ctx.cfg.color(SECTION, "graph_bg", "#0b0d11"),
    );
    ui.painter().rect_stroke(
        rect,
        CornerRadius::same(6),
        Stroke::new(1.0_f32, ctx.cfg.color(SECTION, "cell_border", "#ffffff20")),
        egui::StrokeKind::Inside,
    );
    let grid = ctx.cfg.color(SECTION, "grid", "#ffffff14");
    for fr in [0.0_f32, 0.5, 1.0] {
        let y = rect.bottom() - fr * rect.height();
        ui.painter().line_segment(
            [Pos2::new(rect.left(), y), Pos2::new(rect.right(), y)],
            Stroke::new(1.0_f32, grid),
        );
    }

    let h = hist().lock().unwrap_or_else(|e| e.into_inner());
    if h.samples.len() < 2 {
        return;
    }
    let window = ctx.cfg.f64_key(SECTION, "history_seconds", 6.0);
    // Wall clock (not last sample) so the trace keeps scrolling while pedals are held.
    let now = h.clock.elapsed().as_secs_f64();
    let lw = ctx.cfg.f64_key(SECTION, "line_width", 2.4) as f32;
    let to_pt = |t: f64, frac: f32| -> Pos2 {
        let x = rect.right() - ((now - t) / window).min(1.0) as f32 * rect.width();
        let y = rect.bottom() - frac * rect.height();
        Pos2::new(x, y)
    };

    let thr = brake_threshold(ctx);
    if thr > 0.0 {
        let y = rect.bottom() - thr * rect.height();
        ui.painter().line_segment(
            [Pos2::new(rect.left(), y), Pos2::new(rect.right(), y)],
            Stroke::new(1.4_f32, ctx.cfg.color(SECTION, "threshold", "#ffffff66")),
        );
    }

    let draw_chan = |ui: &mut Ui, di: usize, color: Color32| {
        let pts: Vec<Pos2> = h
            .samples
            .iter()
            .map(|s| {
                let v = match di {
                    1 => s.1,
                    2 => s.2,
                    3 => s.3,
                    4 => s.4,
                    _ => 0.0,
                };
                to_pt(s.0, v)
            })
            .collect();
        if pts.len() >= 2 {
            ui.painter().add(Shape::line(pts, Stroke::new(lw, color)));
        }
    };

    if ctx.cfg.bool_key(SECTION, "show_throttle", true) {
        draw_chan(ui, 1, ctx.cfg.color(SECTION, "throttle", "#46df7a"));
    }
    if ctx.cfg.bool_key(SECTION, "show_clutch", false) {
        draw_chan(ui, 3, ctx.cfg.color(SECTION, "clutch", "#3aa0ff"));
    }
    if ctx.cfg.bool_key(SECTION, "show_steering", false) {
        draw_chan(ui, 4, ctx.cfg.color(SECTION, "steering", "#c08bff"));
    }

    // Brake segment-by-segment for ABS / over colors.
    if ctx.cfg.bool_key(SECTION, "show_brake", true) {
        let mut prev: Option<Pos2> = None;
        for s in h.samples.iter() {
            let cur = to_pt(s.0, s.2);
            if let Some(p0) = prev {
                let col = brake_color(ctx, s.2, s.5 > 0.5, thr);
                ui.painter().line_segment([p0, cur], Stroke::new(lw, col));
            }
            prev = Some(cur);
        }
    }

    if ctx.cfg.bool_key(SECTION, "show_shift_markers", false) {
        let mut prev_g: Option<i32> = None;
        for s in h.samples.iter() {
            if let Some(pg) = prev_g {
                if s.6 != pg {
                    let pt = to_pt(s.0, 0.5);
                    ui.painter().line_segment(
                        [
                            Pos2::new(pt.x, rect.top() + 2.0),
                            Pos2::new(pt.x, rect.bottom() - 2.0),
                        ],
                        Stroke::new(1.2_f32, ctx.cfg.color(SECTION, "text", "#f4f6f8")),
                    );
                }
            }
            prev_g = Some(s.6);
        }
    }
}

fn draw_bars(
    ui: &mut Ui,
    ctx: &mut WidgetCtx<'_>,
    rect: Rect,
    bw: f32,
    bgap: f32,
    chans: &[(usize, &str)],
) {
    let h = hist().lock().unwrap_or_else(|e| e.into_inner());
    let latest = h
        .samples
        .back()
        .copied()
        .unwrap_or((0.0, 0.0, 0.0, 0.0, 0.5, 0.0, 0));
    drop(h);
    let abs_on = latest.5 > 0.5;
    let thr = brake_threshold(ctx);

    let id = egui::Id::new("inputs_bar_anim");
    let mut anim = ui
        .ctx()
        .data_mut(|d| d.get_temp::<BarAnim>(id).unwrap_or_default());
    let dt = anim_dt(ctx.mono_secs, &mut anim.last_secs);
    anim.thr = ease(anim.thr, latest.1, dt, BAR_EASE_TAU);
    anim.brk = ease(anim.brk, latest.2, dt, BAR_EASE_TAU);
    anim.clt = ease(anim.clt, latest.3, dt, BAR_EASE_TAU);
    let animating = still_easing(anim.thr, latest.1, 0.005)
        || still_easing(anim.brk, latest.2, 0.005)
        || still_easing(anim.clt, latest.3, 0.005);
    if animating {
        *ctx.panel_animating = true;
        ui.ctx()
            .request_repaint_after(std::time::Duration::from_millis(1));
    }
    let eased = (anim.thr, anim.brk, anim.clt);
    ui.ctx().data_mut(|d| d.insert_temp(id, anim));

    let label_h = (rect.height() * 0.16).max(10.0);
    let track_top = rect.top() + label_h;
    let track_h = rect.height() - label_h;
    let mut x = rect.left();
    for &(di, colk) in chans {
        let val = match di {
            1 => eased.0,
            2 => eased.1,
            3 => eased.2,
            _ => 0.0,
        }
        .clamp(0.0, 1.0);
        let fill = if di == 2 {
            brake_color(ctx, val, abs_on, thr)
        } else {
            ctx.cfg.color(SECTION, colk, "#46df7a")
        };
        let track = Rect::from_min_size(Pos2::new(x, track_top), Vec2::new(bw, track_h));
        let r = (bw * 0.4) as u8;
        ui.painter().rect_filled(
            track,
            CornerRadius::same(r),
            ctx.cfg.color(SECTION, "bar_track", "#262b34"),
        );
        let fh = val * track_h;
        if fh > 0.5 {
            ui.painter().rect_filled(
                Rect::from_min_size(Pos2::new(x, track_top + track_h - fh), Vec2::new(bw, fh)),
                CornerRadius::same(r),
                fill,
            );
        }
        label(
            ui,
            Pos2::new(x + bw * 0.5, rect.top() + label_h * 0.5),
            Align2::CENTER_CENTER,
            &format!("{:.0}", val * 100.0),
            (rect.height() * 0.14).max(8.0),
            ctx.cfg.color(SECTION, "text", "#f4f6f8"),
            true,
        );
        x += bw + bgap;
    }
}

fn draw_gauge(ui: &mut Ui, ctx: &WidgetCtx<'_>, rect: Rect) {
    let cx = rect.center().x;
    let cy = rect.center().y;
    let rad = rect.width() * 0.5;
    let ring = rect.shrink(rad * 0.06);
    ui.painter().circle_filled(
        ring.center(),
        ring.width() * 0.5,
        ctx.cfg.color(SECTION, "gauge_bg", "#0b0d11"),
    );
    ui.painter().circle_stroke(
        ring.center(),
        ring.width() * 0.5,
        Stroke::new(
            (rad * 0.10).max(2.0),
            ctx.cfg.color(SECTION, "gauge_ring", "#333a42"),
        ),
    );
    label(
        ui,
        Pos2::new(cx, cy - rad * 0.15),
        Align2::CENTER_CENTER,
        &gear_str(ctx.frame.gear),
        rad * 0.85,
        ctx.cfg.color(SECTION, "text", "#f4f6f8"),
        true,
    );
    label(
        ui,
        Pos2::new(cx, cy + rad * 0.28),
        Align2::CENTER_CENTER,
        ctx.cfg.speed_unit(),
        rad * 0.22,
        ctx.cfg.color(SECTION, "muted", "#8b93a1"),
        true,
    );
    let spd = ctx.cfg.conv_speed(ctx.frame.speed_mps);
    label(
        ui,
        Pos2::new(cx, cy + rad * 0.55),
        Align2::CENTER_CENTER,
        &format!("{:.0}", spd),
        rad * 0.30,
        ctx.cfg.color(SECTION, "text", "#f4f6f8"),
        true,
    );
}
